### PR TITLE
CRM-20913 - Add separate option group for pledge status

### PR DESCRIPTION
**Pull request by jitendrapurohit**
Tagging as wip since I haven't yet performed any initial testing for this or written any unit test..

1. Creates a new option_group for pledge similar to contribution_status except `Failed`, `Chargeback`, `Pending Refund`, `Partially paid` and `Refunded`.
2. ~~Remove `In Progress` value from contribution status.~~ - reverted this as `In Progress` gets set for recurring contribution.

Comments welcome @eileenmcnaughton

---

 * [CRM-20913: Separate pledge statuses from contribution statuses](https://issues.civicrm.org/jira/browse/CRM-20913)

### CODE REVIEW

File: CRM/Pledge/BAO/Query.php [hunk 1]
```diff
@@ -405,7 +405,7 @@ public static function from($name, $mode, $side) {
         break;
 
       case 'pledge_status':
-        $from .= " $side JOIN civicrm_option_group option_group_pledge_status ON (option_group_pledge_status.name = 'contribution_status')";
+        $from .= " $side JOIN civicrm_option_group option_group_pledge_status ON (option_group_pledge_status.name = {$name})";
```
Comment (eileenmcnaughton) [inline CRM/Pledge/BAO/Query.php]: Does using $name add anything here - seems it would always be that

**Comment by eileenmcnaughton** [comment]
Not having tested it either, the code looks about right - looks like Jenkins has been busy testing it :-)

File: CRM/Pledge/BAO/Query.php (hunk 1 above)
Comment (jitendrapurohit) [inline CRM/Pledge/BAO/Query.php]: `$name` is always `pledge_status`. But, just for readability, I'll update this to text instead of variable.

**Comment by jitendrapurohit** [comment]
Tested the following scenarios with the PR and worked fine. 

1. Online pledge contribution with pay later(pledge_status = pending) and CC(pledge status = inprogress for multiple installments).
2. Backend pledge creation and recording payments(pledge status after each payments looked fine).
3. Cancellation of pledge.
4. Search related to Pledge in Advanced Search, Find Pledge with various status options and check built qill message.
5. Pledge Details Report with filtering on status and other options.

To-do - Test Upgrade code to assert whether new option group gets created after upgradation.

**Comment by jitendrapurohit** [comment]
Tested and confirmed `pledge_status` gets correctly created after upgrade. Removing wip tag as I've done writing unit test and manual testing from my side.

**Comment by eileenmcnaughton** [comment]
@monishdeb if you are able to take a look at this it should simplify some of the other contribution things

**Comment by jitendrapurohit** [comment]
Note: I haven't yet pushed the generated data changes, as PR is more likely to get stale due to them.

**Comment by monishdeb** [comment]
Tested the PR, working as expected. @jitendrapurohit can you add the generated data changes. So I could merge then?

**Comment by jitendrapurohit** [comment]
Done @monishdeb 👍

**Comment by monishdeb** [comment]
Tested all the five scenarios/fixes mentioned above https://github.com/civicrm/civicrm-core/pull/10737#issuecomment-318002173 . Working fine as expected

File: xml/templates/civicrm_data.tpl [hunk 2]
```diff
@@ -1025,7 +1027,14 @@ VALUES
...
-   (@option_group_id_date_filter, '{ts escape="sql"}From end of previous calendar year{/ts}', 'greater_previous.year', 'greater_previous.year', NULL, NULL, NULL,62, NULL, 0, 0, 1, NULL, NULL, NULL);
+   (@option_group_id_date_filter, '{ts escape="sql"}From end of previous calendar year{/ts}', 'greater_previous.year', 'greater_previous.year', NULL, NULL, NULL,62, NULL, 0, 0, 1, NULL, NULL, NULL),
+
+-- Pledge Status
+  (@option_group_id_ps, '{ts escape="sql"}Completed{/ts}'  , 1, 'Completed'  , NULL, 0, NULL, 1, NULL, 0, 1, 1, NULL, NULL, NULL),
+  (@option_group_id_ps, '{ts escape="sql"}Pending{/ts}'    , 2, 'Pending'    , NULL, 0, NULL, 2, NULL, 0, 1, 1, NULL, NULL, NULL),
+  (@option_group_id_ps, '{ts escape="sql"}Cancelled{/ts}'  , 3, 'Cancelled'  , NULL, 0, NULL, 3, NULL, 0, 1, 1, NULL, NULL, NULL),
+  (@option_group_id_ps, '{ts escape="sql"}In Progress{/ts}', 4, 'In Progress', NULL, 0, NULL, 4, NULL, 0, 1, 1, NULL, NULL, NULL),
+  (@option_group_id_ps, '{ts escape="sql"}Overdue{/ts}'    , 5, 'Overdue'    , NULL, 0, NULL, 5, NULL, 0, 1, 1, NULL, NULL, NULL);
```
Comment (monishdeb) [inline xml/templates/civicrm_data.tpl]: @jitendrapurohit @eileenmcnaughton I forgot to mention, do we need to delete ```In Progress``` and ```Overdue``` status from Contribution Status's option values too?

**Comment by eileenmcnaughton** [comment]
Well done guys - I think @jitendrapurohit said 'In Progress' is needed for contribution recur (which suggests to me it should have it's own set too)

File: xml/templates/civicrm_data.tpl (hunk 2 above)
Comment (jitendrapurohit) [inline xml/templates/civicrm_data.tpl]: These are pledge status where `In Progress` and `Overdue` status are required. I haven't removed any existing status in this PR. See [point 2 in first comment](https://github.com/civicrm/civicrm-core/pull/10737#issue-245051750) where I initially removed `Inprogress` from contribution status but it looks like some unit test do want to set this status for recur contribution - Eg. - [ContributionTest](https://github.com/civicrm/civicrm-core/blob/master/tests/phpunit/api/v3/ContributionTest.php#L2587)  and [ContributionPageTest](https://github.com/civicrm/civicrm-core/blob/master/tests/phpunit/api/v3/ContributionPageTest.php#L1468).

**Pull request merged**